Fix reload gunicorn workers

AVMusorin opened the pull request:
Since gunicorn can't reload a new code if starts with ``--preload`` setting, we need to check ``reload_on_plugin_change`` before set it up.

Gunicorn can't reload a new code because the code is preloaded into the master process and worker are launched with ``fork``, they will still have the old code.

potiuk reviewed (dismissed):
I think there was a good reason we added `--preload`, https://github.com/apache/airflow/pull/27297/files and it was to remove lots of integrity warnings generated by parallel gunicorns trying to update the database in parallell and generating race conditions.

This cahnge means that when "reload_on_plugin_change" is turned on, we will start getting those warnings, unless the root cause of the problem has been fixed (I think it hasn't been fixed).

So I think this change only makes sense to be implemented if we find another way to prevent the integrity errors. For example adding some locking strategy for the DB operations that generated integrity warnings - without it, it will be a regression.

AVMusorin committed:
Toggle gunicorn --preload with reload_on_plugin_change

Since gunicorn can't reload a new code if starts with ``--preload``
setting, we need to check ``reload_on_plugin_change`` before set it up.

Gunicorn can't reload a new code because the code is preloaded into the
master process and worker are launched with ``fork``, they will still have
the old code.

AVMusorin committed:
added warning message

ephraimbuddy reviewed (changes requested):
To prevent accidental merge

jedcunningham commented on airflow/cli/commands/webserver_command.py:
```diff
@@ -422,11 +422,18 @@ def webserver(args):
+                "reload_on_plugin_change prevent running gunicorn with --preload option. "
+                "With the preload option, the app is loaded before the workers are forked, and each worker "
+                "will then have a copy of the app. It may cause IntegrityError during webserver startup, so "
+                "avoid using reload_on_plugin_change in production."
```
```suggestion
                "will then have a copy of the app. reload_on_plugin_change may cause IntegrityErrors during "
                "webserver startup, so avoid using it in production."
```

jedcunningham commented on airflow/cli/commands/webserver_command.py:
```diff
@@ -422,11 +422,18 @@ def webserver(args):
+            run_args += ["--preload"]
+        if reload_on_plugin_change:
+            log.warning(
+                "reload_on_plugin_change prevent running gunicorn with --preload option. "
```
```suggestion
                "reload_on_plugin_change prevents running gunicorn with the --preload option. "
```

nit

jedcunningham commented on airflow/config_templates/config.yml:
```diff
@@ -1445,7 +1445,8 @@ webserver:
         If set to True, Airflow will track files in plugins_folder directory. When it detects changes,
-        then reload the gunicorn.
+        then reload the gunicorn. If set to True, gunicorn starts without ``--preload`` setting and uses
+        more memory.
```
```suggestion
        then reload the gunicorn. If set to True, gunicorn starts without ``--preload``, which
        is slower and uses more memory.
```

AVMusorin committed:
Improve warning message

Co-authored-by: Jed Cunningham <66968678+jedcunningham@users.noreply.github.com>

AVMusorin committed:
Improve warning message

Co-authored-by: Jed Cunningham <66968678+jedcunningham@users.noreply.github.com>

AVMusorin committed:
Improve warning message

Co-authored-by: Jed Cunningham <66968678+jedcunningham@users.noreply.github.com>

uranusjr committed:
Minor tweak of warning message

uranusjr committed:
Mention prod issue in config description

uranusjr reviewed (approved):
I added a couple of commits to refine the message. The warning messag especially has some wording issues.
